noir_stdlib/src/meta/expr.nr

```diff
@@ -110,4 +110,51 @@ impl Expr {
     // docs:start:is_continue
     fn is_continue(self) -> bool {}
     // docs:end:is_continue
+
+    fn map<Env>(self, f: fn[Env](Expr) -> Option<Expr>) -> Expr {
+        let result = f(self);
+        let result = result.or_else(|| map_if(self, f));
+        let result = result.or_else(|| map_block(self, f));
+        // TODO: check the remaining possibilities...
```

Comment:
Remaining TODO

Also, please add docs for these new functions. I'd guess that `map` here is something like "attempt to run the function on the whole expression and then each option for sub-expression in [??] order" so perhaps `try_map_subexpression` or a similar name would avoid confusion with collection-style map functions?

Also, the ordering of the sub-expressions seems a bit error prone to me:
- As defined, adding support for different expressions in `f` could change which sub-expression `f` gets executed on
  + I.e., the ordering of which case is tried first seems both significant and unpredictable.
- My favorite alternative is to run each case, pick the first successful case as you're doing here, and then assert that exactly one case succeeded.

Reply:
I _think_ there's no issue in how it's currently done:
- First we call `f(self)`. If that returns `Some` we'll return that (all of the other `or_else` won't call their lambda)
- Otherwise it's none so we check which expression we are. Now here only one of the next `or_else` lambdas will be called, because the expression can only be one of `if`, `array`, `block`, etc. And for the one that it returns `Some`, that'll be the returned value.

So... I think the order in which we evaluate these doesn't matter?

Reply:
I added some docs... I'm still not sure about the docs for `map`. Maybe @jfecher could come up with a better description? (he had the idea of this method and its signature)